Static method to convert a Map to a valid JSON string.

jsfan3 opened the pull request:
I noted that the CodenameOne API is missing of a method to convert a Map<String,Object> to a JSON string, that's why I created a static method to convert the given Map<String,Object> to a valid JSON representation. The values allowed types in the Map<String,Object> are: Number, String, Boolean, List, Map or null. Limited whitespace is inserted be make the resulting JSON string more readable. Simple and complex examples of usage are in the javadoc. I tested the generated JSON strings in a JSON validator. I hope that my proposed method will be helpful.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+     * JSON string fails.
+     * @author Francesco Galgani
+     */
+    public static String mapToJson(Map<String, Object> map) throws Exception {
```
- The argument should be `Map<String, ?>`.
- Throwing a checked exception is a big pain for such utils.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+            } else if (entry.getValue() instanceof Number) {
+                String value;
+                if (entry.getValue() instanceof Byte) {
+                    value = ((Byte) entry.getValue()).toString();
```
This cast and all following are needless. You're just testing if it's a `Number` of a kind you accept. Not sure, if it's worth it. What about `Short` or `BigInteger`?

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+                } else if (entry.getValue() instanceof Integer) {
+                    value = ((Integer) entry.getValue()).toString();
+                } else if (entry.getValue() instanceof Long) {
+                    value = ((Long) entry.getValue()).toString();
```
As JS has nothing but `Double` and therefore `Long` can't be represented exactly as number, it's common to offer an option serializing `Long` to `String`.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+                } else if (entry.getValue() instanceof Long) {
+                    value = ((Long) entry.getValue()).toString();
+                } else {
+                    throw new Exception("Cast error in converting map to json");
```
Such a message isn't helpful. You need the key and the value.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+                value.append("[");
+                if (!myList.isEmpty() && (myList.get(0) instanceof String)) {
+                    for (int i = 0; i < myList.size(); i++) {
+                        String item = (String) (myList.get(i));
```
- This fails badly for heterogeneous lists.
- It's needlessly repetitive and not very general, what about list of lists? It needs recursion.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+                } else if (!myList.isEmpty() && myList.get(0) instanceof Map) {
+                    for (int i = 0; i < myList.size(); i++) {
+                        String item = mapToJson((Map) (myList.get(i)));
+                        List<String> lines = StringUtil.tokenize(item, '\n');
```
Inefficient. Use a helper method like `appendTo(StringBuilder result, Map<String, ?>, int depth)`.

And another one for lists.

Maaartinus commented on CodenameOne/src/com/codename1/io/JSONParser.java:
```diff
@@ -632,4 +634,270 @@ public void keyValue(String key, String value) {
+                value.append("]");
+                json.append(value.toString());
+            } else if (entry.getValue() instanceof Map) {
+                String value = mapToJson((Map) entry.getValue());
```
What if the keys aren't strings?

codenameone commented:
You can replace all of that code with:

````java
return Result.fromContent(map).toString();
````

It's not well documented but we use it here and there.

jsfan3 commented:
Thank you "Maaartinus" for the very precise comments.

Thank you "codenameone" (are you Shai?) for solving the problem with one line: I tested your solution against the same Maps that I used with my code and the obtained results are equivalent.
In this new commit, I removed all my code, I inserted your one-line solution, I updated the javadoc with the example outputs resulting from Result.fromContent(map).toString();

Maybe adding this method and its javadoc can be helpfull, because otherwise it's not so intuitive how to produce and convert a Map to a Json string.

codenameone commented:
Thanks.
Yes I agree this method signature would help document that more effectively.
